### PR TITLE
Memory benchmark for showing large images in succession

### PR DESCRIPTION
**Pull request by dnfield**
Benchmark for https://github.com/flutter/flutter/issues/56482

This will be improved when the following finish landing:

- https://github.com/flutter/engine/pull/20754 
- https://github.com/flutter/flutter/pull/64582

Current behavior observed is memory growth until a GC is finally performed after multiple iterations of changing the image. Expected behavior will be that a GC is performed at least once after a large image is disposed of.

### CODE REVIEW

File: dev/benchmarks/macrobenchmarks/lib/src/large_image_changer.dart [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+    );
+    _timer?.cancel();
+    _timer = Timer.periodic(const Duration(seconds: 3), (Timer timer) {
+      currentImage.evict().then((_) {
+        setState(() {
+          if (imageIndex > 5) {
+            imageIndex = 0;
+          } else {
+            imageIndex += 1;
+          }
```
Comment (liyuqian) [inline dev/benchmarks/macrobenchmarks/lib/src/large_image_changer.dart:37]: Nit: line 33-37 can be simplified as `imageIndex = (imageIndex + 1) % 6`.

File: dev/devicelab/bin/tasks/large_image_changer_perf.dart [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+import 'package:flutter_devicelab/framework/adb.dart';
+import 'package:flutter_devicelab/framework/framework.dart';
+import 'package:flutter_devicelab/framework/utils.dart';
+import 'package:flutter_devicelab/tasks/perf_tests.dart';
+
+const String kPackageName = 'com.example.macrobenchmarks';
+const String kActivityName = 'com.example.macrobenchmarks.MainActivity';
+
+class LargeImageChangerPerfTest extends MemoryTest {
```
Comment (liyuqian) [inline dev/devicelab/bin/tasks/large_image_changer_perf.dart:15]: I wonder if more memory info (e.g., peak memory, average, percentile, etc.) from `DevToolsMemoryTest` (see, e.g., [this example][1]) or Timeline (see, e.g., [profiling_summarizer][2] and https://github.com/flutter/engine/pull/18516) would be helpful here? The DevTools works for Android and the Timeline works for iOS.

[1]: https://github.com/flutter/flutter/blob/master/dev/devicelab/bin/tasks/complex_layout_scroll_perf__devtools_memory.dart

[2]: https://github.com/flutter/flutter/blob/master/packages/flutter_driver/lib/src/driver/profiling_summarizer.dart#L79

**Comment by dnfield** [comment]
Updated to use a DevTools memory test.

File: dev/benchmarks/macrobenchmarks/lib/src/large_image_changer.dart (hunk 1 above)
Comment (dnfield) [reply dev/benchmarks/macrobenchmarks/lib/src/large_image_changer.dart]: Done

**Comment by jonahwilliams** [comment]
Is it feasible to measure CPU usage as well?

**Comment by dnfield** [comment]
@liyuqian might have more context on that.

What CPU benchmarks do you think this would capture that we're not already capturing though?

**Comment by jonahwilliams** [comment]
specifically the cost of increased GCs to reduce image memory usage

**Comment by dnfield** [comment]
Added an iOS variant that does CPU as well.

File: dev/benchmarks/macrobenchmarks/test_driver/large_image_changer_test.dart [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    if (timeline != null) {
+      final TimelineSummary summary = TimelineSummary.summarize(timeline);
+      await summary.writeSummaryToFile(fileName, pretty: true);
+      await summary.writeTimelineToFile(fileName, pretty: true);
+    }
+
+    await driver.close();
+  });
+
+  // test('Measure CPU/GPU/Memory', () async {
```
Comment (jonahwilliams) [inline dev/benchmarks/macrobenchmarks/test_driver/large_image_changer_test.dart:46]: commented out code?

File: dev/devicelab/lib/tasks/perf_tests.dart [hunk 4]
```diff
@@ -1125,11 +1125,15 @@ class DevToolsMemoryTest {
...
           final RegExpMatch match = RegExp(r'An Observatory debugger and profiler on .+ is available at: ((http|//)[a-zA-Z0-9:/=_\-\.\[\]]+)').firstMatch(line);
-          if (match != null) {
+          if (match != null && !observatoryUri.isCompleted) {
             observatoryUri.complete(match[1]);
             _observatoryUri = match[1];
           }
-        }, onDone: () { observatoryUri.complete(null); });
+        }, onDone: () {
+          if (!observatoryUri.isCompleted) {
+            observatoryUri.complete(null);
```
Comment (jonahwilliams) [inline dev/devicelab/lib/tasks/perf_tests.dart:1134]: unless you need the `null`, this can be `observatoryUri.complete()`

**Review by jonahwilliams — APPROVED** [review]
LGTM

File: dev/devicelab/lib/tasks/perf_tests.dart (hunk 4 above)
Comment (dnfield) [reply dev/devicelab/lib/tasks/perf_tests.dart]: Done

File: dev/benchmarks/macrobenchmarks/test_driver/large_image_changer_test.dart (hunk 3 above)
Comment (dnfield) [reply dev/benchmarks/macrobenchmarks/test_driver/large_image_changer_test.dart]: Done

**Review by liyuqian — APPROVED** [review]
LGTM

**Pull request merged**